Decorator namespace key

pshafer-als opened the pull request:
## What do these changes do?

### Keys with namespaces now work for decorators get/set
* Include build_key(key, namespace) in decorators.cached:
    * get_from_cache()
    * set_in_cache()

## Are there changes in behavior for the user?
* Keys with namespaces now work for the @cached() decorator

## Related issue number
* https://github.com/aio-libs/aiocache/issues/569

## Checklist

- [X ] I think the code is well written
- [X ] Unit tests for the changes exist
- [X ] Documentation reflects the changes
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` (e.g. `588.bugfix`)
  * if you don't have an `issue_id` change it to the pr id after creating the PR
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: `Fix issue with non-ascii contents in doctest text files.`

Dreamsorcerer commented on aiocache/decorators.py:
```diff
@@ -134,16 +134,18 @@ def _key_from_args(self, func, args, kwargs):
+    async def get_from_cache(self, key):
         try:
-            return await self.cache.get(key)
+            namespace = self._kwargs.get("namespace", None)
```
Presumably, this shouldn't cause an exception, so move it before the `try`.

Dreamsorcerer commented on aiocache/decorators.py:
```diff
@@ -134,16 +134,18 @@ def _key_from_args(self, func, args, kwargs):
     async def set_in_cache(self, key, value):
         try:
-            await self.cache.set(key, value, ttl=self.ttl)
+            namespace = self._kwargs.get("namespace", None)
```
Same

Dreamsorcerer commented on aiocache/decorators.py:
```diff
@@ -205,7 +207,12 @@ async def decorator(self, f, *args, **kwargs):
 
 def _get_cache(cache=Cache.MEMORY, serializer=None, plugins=None, **cache_kwargs):
-    return Cache(cache, serializer=serializer, plugins=plugins, **cache_kwargs)
+    return Cache(
```
Looks unrelated (maybe I need to tweak a Black config or something...).

pshafer-als replied:
I've deleted space characters at ends of lines here.

Dreamsorcerer commented on tests/acceptance/test_decorators.py:
```diff
@@ -48,6 +48,35 @@ async def fn(self, a, b=2):
+        decorated_fn = cached(fn, namespace=None)
+        
+        await fn()
+        assert await cache.exists(
+            decorated_fn.get_cache_key(fn, args=(), kwargs={})
+            ) is True
```
```suggestion
        key = decorated_fn.get_cache_key(fn, args=(), kwargs={})
        assert await cache.exists(key) is True
```

Dreamsorcerer commented on tests/acceptance/test_decorators.py:
```diff
@@ -48,6 +48,35 @@ async def fn(self, a, b=2):
+        decorated_ns_fn = cached(ns_fn, namespace=key_prefix)
+        
+        await ns_fn()
+        assert await cache.exists(
+            decorated_ns_fn.get_cache_key(ns_fn, args=(), kwargs={})
+            ) is True
+        assert decorated_ns_fn.get_cache_key(
+            ns_fn, args=(), kwargs={},
+            ).startswith(key_prefix)
```
```suggestion
        key = decorated_ns_fn.get_cache_key(ns_fn, args=(), kwargs={})
        assert await cache.exists(key) is True
        assert key.startswith(key_prefix)
```

Dreamsorcerer commented:
Tests are failing.

pshafer-als commented:
> Tests are failing.

Noted.  Changes are in progress now...I'm planning to push the corrections within the next hour.

Dreamsorcerer commented:
Nope, still failing. You're not testing in 3.7 are you? Because only half the tests run on 3.7.

pshafer-als commented:
I've now updated all decorators to use the namespace if provided.

To make this work, I updated `TestCachedStampede.test_cached_stampede` to use the namespace in the call signature of the mock `cache.get()` and `cache.set()`.

This is the intended behavior after the changes I made in this branch to the `decorators.cached`. But I don't know whether there was a good reason to explicitly exclude the namespace in this test previously.

I made similar changes in tests/ut/ so that `mock_cache` uses `namespace="test"`.

Dreamsorcerer commented on aiocache/decorators.py:
```diff
@@ -75,6 +79,7 @@ def __init__(
 
         self._cache = cache
         self._serializer = serializer
+        self._namespace = namespace
```
In some places, you've used `self.cache.namespace`, others `self._namespace`, do we need this extra attribute?

pshafer-als replied:
Probably not.  Comments below suggest that it might not be needed after __init__() completes.

Dreamsorcerer commented on aiocache/decorators.py:
```diff
@@ -134,18 +140,21 @@ def _key_from_args(self, func, args, kwargs):
+        namespace = self.cache.namespace
         try:
-            return await self.cache.get(key)
+            return await self.cache.get(key, namespace=namespace)
```
Isn't the cache already using it's own namespace?

Dreamsorcerer replied:
It looks to me like the build_key method is expected to use `self.namespace` if `None`:
https://github.com/pshafer-als/aiocache/blob/decorator-namespace-key/aiocache/base.py#L499

Dreamsorcerer replied:
I don't see a single example of how to use a custom build_key, so maybe it'd make more sense if all the calls to `self.build_key()` did `namespace=namespace or self.namespace`. Then it's more likely to work as expected.

I think this is potentially backwards-incompatible, so I'd prefer if we did this change in a later PR. Can we get this working without passing the namespace in? Even if a couple of tests need to have xfail on them until that change is made.

pshafer-als replied:
Yes, that was on my mind also.  Some of this is remnant of a shift in strategy to better use the underlying case, rather than implement namespace functionality directly in the decorator.

I'll clean this up and see how the test perform.

Dreamsorcerer commented on aiocache/decorators.py:
```diff
@@ -38,8 +38,11 @@ class cached:
+        the backend. Default is None
     :param key_builder: Callable that allows to build the function dynamically. It receives
         the function plus same args and kwargs passed to the function.
+        This behavior is necessarily different than ``BaseCache.build_key()``
```
I'm looking at the code, and not really seeing the point of this `key_builder`. Maybe we should just remove it and use the cache methods (later)...

pshafer-als replied:
This `key_builder` in the decorators is necessary to make a cache key from the call signature of the function/coroutine that is being decorated.  It's the only place where the "raw" key gets generated. It accepts the callable object and the `args` + `kwargs` from that callable.

Whereas the `key_builder` supplied to `BaseCache` (which becomes `BaseCache.build_key`) accepts only two parameters: `key` and `namespace`, and it concatenates them by default. The "raw" key must already exist.

Perhaps these two functionalities of theses parameters should have separate names, rather than both being designated as `key_builder`?

Dreamsorcerer replied:
OK, let's leave it alone for now then. Can probably do with some better documentation or something later though.

Dreamsorcerer commented:
Looks like we have what we need to make a new release now. Think you can finish this off in the next couple of days?

pshafer-als commented:
> Looks like we have what we need to make a new release now. Think you can finish this off in the next couple of days?

Yes, I can make that work.

pshafer-als commented:
In reference to [this comment](https://github.com/aio-libs/aiocache/issues/569#issuecomment-1376107644) in Issue #569, here is a summary of the responses included in this PR.

1.  N/A
2. ...
    a. See examples/alt_key_builder.py and updates to examples/cached_decorator.py
    b. `BaseCache.build_key` uses `namespace` argument if provided, otherwise it uses `self.namespace`.
    c. Cache locks use `client.build_key` rather than `client._build_key`.
3. ...
    a. _Deferred_. This might get submitted as a separate issue / PR at a later time.
    b. `namespace` is an explicit parameter to decorator caches.
4. _Deferred_. This might get submitted as a separate issue / PR at a later time.
5. Warn when decorator parameters are unused because alias takes precedence.
6. _Deferred_. This might get submitted as a separate issue / PR at a later time.
7. N/A

pshafer-als commented:
@Dreamsorcerer  I think this is ready to go, but let me know if you see additional changes needed.

Dreamsorcerer commented on examples/alt_key_builder.py:
```diff
@@ -0,0 +1,293 @@
+"""alt_key_builder.py
```
This file feels a little too verbose/dense for an example, but I'll merge it for now and look at it again in a later release.

pshafer-als commented on tests/acceptance/test_lock.py:
```diff
@@ -97,6 +115,26 @@ class TestRedisRedLock:
         return RedLock(redis_cache, Keys.KEY, 20)
 
+    @pytest.fixture
+    def set_key_builder(self, redis_cache):
```
DONE

pshafer-als commented on examples/cached_decorator.py:
```diff
@@ -16,12 +17,56 @@ async def cached_call():
+)
+
+
+async def test_cached_alias_build_key():
```
DONE

pshafer-als reviewed (commented):
All comments appear to be resolved

pshafer-als commented:
@Dreamsorcerer Please have another look, and let me know if all is good.

Dreamsorcerer commented:
Now, if you're happy to do some more, it would be great if you can create a PR which reverts all the extra namespace conversions in here and does it something like this:
```
def __init__(self, key_builder):
    self._build_key = key_builder or lambda k, ns: k if ns is None else "{}{}".format(ns, k)

def build_key(self, key, namespace=None):
    return self._build_key(key, namespace or self.namespace)
```

Then we can do it like this for v1 and introduce some backwards incompatible changes.

Should also be able to fold in the `_ensure_key()` logic and remove that function. Plus, MemcachedBackend can do the `.encode()` in `build_key()` as well.

In other words, the key_builder should not need to know about any of these edge cases, it should just receive a `str` and return a `str`.

pshafer-als commented:
Nice!  I like those ideas.  

I should have some time over the next couple weeks to give this a try.

Dreamsorcerer commented:
@pshafer-als Any chance you'll get round to that soon?

pshafer-als commented:
That's a fair question.  Realistically, I think it will be ~2 weeks before I can secure sufficient time to work on this in earnest.  I think it will be a few hours of work, but probably spread over a 1-2 week period.

If you need to move on, I understand. Otherwise, I'm still interested.

Dreamsorcerer commented:
That's good, just making sure the task doesn't get lost.

pshafer-als commented:
Started a [branch for this work](https://github.com/pshafer-als/aiocache/blob/simplify-cache-namespace/aiocache/base.py), and a [new thread for discussion](https://github.com/aio-libs/aiocache/issues/669).
